File: src/features/test-card/test-card.presenter.ts
import { color } from '../../infrastructure/terminal/terminal-style'
import type { TerminalOutput } from '../../app/ports'
import type { Config } from '../../infrastructure/config/config.models'
import {
  TEST_CARD_SYSTEM_PROMPT,
  workItemText,
  type TestCardPreparation
} from './test-card.models'

export class TestCardPresenter {
  constructor(private readonly output: TerminalOutput) {}

  intro(branch: string): void {
    this.output.write(`${color('magenta', `┌ prt · Test Case ${branch}`)}\n│`)
  }

  outro(message: string): void {
    this.output.write(color('magenta', `└ ${message}`))
  }

  showDryRun(config: Config, prompt: string): void {
    const models = config.providers
      .map((provider) => {
        const model =
          provider === 'codex'
            ? config.codexModel
            : provider === 'opencode'
              ? config.opencodeModel
              : config.compatibleModel
        const reasoning =
          provider === 'codex'
            ? config.codexReasoning
            : provider === 'opencode'
              ? config.opencodeReasoning
              : config.compatibleReasoning
        return `${provider}/${model} (thinking ${reasoning})`
      })
      .join(', ')
    this.output.write(`Provider/model: ${models}`)
    this.output.write('\n[SYSTEM]\n')
    this.output.write(TEST_CARD_SYSTEM_PROMPT)
    this.output.write('\n[USER]\n')
    this.output.write(prompt)
  }

  showSummary(
    preparation: TestCardPreparation,
    provider: string,
    model: string,
    title: string,
    body: string,
    options: { areaPath?: string; assignedTo?: string }
  ): void {
    const { context, config } = preparation
    this.output.write(`\nTest Card${context.pullRequest ? ` · PR #${context.pullRequest.pullRequestId}` : ''}`)
    this.output.write(`Provider: ${provider}/${model}`)
    this.output.write(
      `Work Item: #${context.workItem.id} — ${workItemText(context.workItem, 'System.Title')}`
    )
    this.output.write(`AreaPath: ${(options.areaPath ?? config.testAreaPath) || '(não configurado)'}`)
    const assignedTo = options.assignedTo ?? config.testAssignedTo
    if (assignedTo) this.output.write(`Responsável: ${assignedTo}`)
    this.output.write(`\nTítulo: ${title}\n\n${body}`)
  }

  raw(body: string): void {
    this.output.write(body)
  }

  success(message: string): void {
    this.output.write(color('green', `✓ ${message}`))
  }

  info(message: string): void {
    this.output.write(color('cyan', `• ${message}`))
  }
}

File: lib/src/features/test_card/test_card_presenter.dart
import 'package:better_effect/better_effect.dart';

import '../../app/app_effect.dart';
import '../../app/cli_options.dart';
import '../../application/config/config_models.dart';
import '../../application/terminal/terminal_ports.dart';
import 'test_card_models.dart';

abstract interface class TestCardPresenter {
  AppEffect<Unit> intro(String branch);

  AppEffect<Unit> outro(String message);

  AppEffect<Unit> showDryRun(Config config, String prompt);

  AppEffect<Unit> showSummary(
    TestCardPreparation preparation,
    String provider,
    String model,
    String title,
    String body,
    CliOptions options,
  );

  AppEffect<Unit> raw(String body);

  AppEffect<Unit> success(String message);

  AppEffect<Unit> info(String message);
}

final class TestCardPresenterLive implements TestCardPresenter {
  const TestCardPresenterLive();

  @override
  AppEffect<Unit> intro(String branch) => Effect.result((use) {
    use<TerminalOutput>().heading('Test Case', detail: branch);
    return unit;
  });

  @override
  AppEffect<Unit> outro(String message) => _success(message);

  @override
  AppEffect<Unit> showDryRun(Config config, String prompt) =>
      Effect.result((use) {
        final output = use<TerminalOutput>();
        output.heading('Test Case · dry run');
        output.info('Provider/model: ${_models(config)}');
        output.detail('Prompt de sistema');
        output.write(testCardSystemPromptValue);
        output.detail('Prompt de usuário');
        output.write(prompt);
        return unit;
      });

  @override
  AppEffect<Unit> showSummary(
    TestCardPreparation preparation,
    String provider,
    String model,
    String title,
    String body,
    CliOptions options,
  ) => Effect.result((use) {
    final output = use<TerminalOutput>();
    final context = preparation.context;
    output.heading(
      'Test Card${context.pullRequest == null ? '' : ' · PR #${context.pullRequest!.id}'}',
    );
    output.info('Título: $title');
    output.detail('Provider: $provider/$model');
    output.detail(
      'Work Item: #${context.workItem.id} — ${workItemText(context.workItem, 'System.Title')}',
    );
    output.detail(
      'AreaPath: ${(options.areaPath ?? preparation.config.testAreaPath).isEmpty ? '(não configurado)' : options.areaPath ?? preparation.config.testAreaPath}',
    );
    final assigned = options.assignedTo ?? preparation.config.testAssignedTo;
    if (assigned.isNotEmpty) output.detail('Responsável: $assigned');
    output.write('\n$body');
    return unit;
  });

  @override
  AppEffect<Unit> raw(String body) => _write(body);

  @override
  AppEffect<Unit> success(String message) => _success(message);

  @override
  AppEffect<Unit> info(String message) => Effect.result((use) {
    use<TerminalOutput>().info(message);
    return unit;
  });

  AppEffect<Unit> _success(String message) => Effect.result((use) {
    use<TerminalOutput>().success(message);
    return unit;
  });

  AppEffect<Unit> _write(String value) => Effect.result((use) {
    use<TerminalOutput>().write(value);
    return unit;
  });
}

String _models(Config config) => config.providers
    .map((provider) {
      final model = switch (provider) {
        'codex' => config.codexModel,
        'opencode' => config.opencodeModel,
        _ => config.compatibleModel,
      };
      final reasoning = switch (provider) {
        'codex' => config.codexReasoning,
        'opencode' => config.opencodeReasoning,
        _ => config.compatibleReasoning,
      };
      return '$provider/$model (thinking $reasoning)';
    })
    .join(', ');
